### PR TITLE
[Merged by Bors] - Stop registering of protocols after p2p has started

### LINKED ISSUE #571: Don't let protocols register after p2p has started
# Overview / Motivation

Now we allow protocols to register whenever they want. but the assumption is that protocols must be registered before `p2p` has started.

# The Task

block protocols registrations after `p2p` has started. (panic/error)

# Implementation Notes
TODO: Add links to relevant resources, specs, related issues, etc...

# Contribution Guidelines
Important: Issue assignment to developers will be by the order of their application and proficiency level according to the tasks complexity. We will not assign tasks to developers who have'nt introduced themselves on our Gitter [dev channel](https://gitter.im/spacemesh-os/Lobby)

1. Introduce yourself on go-spacemesh [dev chat channel](https://gitter.im/spacemesh-os/Lobby) - ask our team any question you may have about this task
2. Fork branch `develop` to your own repo and work in your repo
3. You must document all methods, enums and types with [godoc comments](https://blog.golang.org/godoc-documenting-go-code)
4. You must write go unit tests for all types and methods when submitting a component, and integration tests if you submit a feature
5. When ready for code review, submit a PR from your repo back to branch `develop`
6. Attach relevant issue to PR

### PR DESCRIPTION
**Pull request by sidsankhe**
## Motivation

Closes #571

## Changes

Added a panic if a protocol tries to register after the p2p has started and removed the `protocolHandlerMutex` because it was no longer needed. 
## Test Plan

Existing tests were modified so that they did not call registration of protocols after the p2p had started
## TODO

- [ ] Explain motivation or link existing issue(s)
- [ ] Test changes and document test plan
- [ ] Update documentation as needed

### CODE REVIEW

File: p2p/switch.go [hunk 1]
```diff
@@ -409,20 +408,22 @@ func (s *Switch) sendMessageImpl(peerPubKey p2pcrypto.PublicKey, protocol string
 
 // RegisterDirectProtocol registers an handler for a direct messaging based protocol.
 func (s *Switch) RegisterDirectProtocol(protocol string) chan service.DirectMessage { // TODO: not used - remove
+	if s.started == 1 {
+		panic("Calling register after the p2p has started")
```
Comment (y0sher) [inline p2p/switch.go]: I would use `log.Panic` here for consistency across the project and for better support for logging panics.

File: p2p/switch.go [hunk 2]
```diff
@@ -409,20 +408,22 @@ func (s *Switch) sendMessageImpl(peerPubKey p2pcrypto.PublicKey, protocol string
...
-	s.protocolHandlerMutex.Lock()
 	s.directProtocolHandlers[protocol] = mchan
-	s.protocolHandlerMutex.Unlock()
 	return mchan
 }
 
 // RegisterGossipProtocol registers an handler for a gossip based protocol. priority must be provided.
 func (s *Switch) RegisterGossipProtocol(protocol string, prio priorityq.Priority) chan service.GossipMessage {
+	if s.started == 1 {
+		panic("Calling register after the p2p has started")
```
Comment (y0sher) [inline p2p/switch.go]: same comment as above

File: p2p/switch.go [hunk 3]
```diff
@@ -485,9 +484,10 @@ func (s *Switch) processMessage(ime net.IncomingMessageEvent) {
 
 // RegisterDirectProtocolWithChannel registers a direct protocol with a given channel. NOTE: eventually should replace RegisterDirectProtocol
 func (s *Switch) RegisterDirectProtocolWithChannel(protocol string, ingressChannel chan service.DirectMessage) chan service.DirectMessage {
-	s.protocolHandlerMutex.Lock()
+	if s.started == 1 {
+		panic("Attempting to register after p2p has started")
```
Comment (y0sher) [inline p2p/switch.go]: same here

Comment (y0sher) [reply p2p/switch.go]: also I think its worth to either unify the message or include some info about what is the origin of the panic

**Review by y0sher — APPROVED** [review]
Looks good!  honestly I thought this will introduce issues with other protocols but I'm happy it went smooth. good job! 
I added minor comments, please address them and merge :).

**Comment by sidsankhe** [comment]
bors merge

**Pull request closed without merging**